File: app/src/main/java/eu/kanade/tachiyomi/data/rest/Release.java
package eu.kanade.tachiyomi.data.rest;

import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Release object
 * Contains information about the latest release
 */
public class Release {
    /**
     * Version name V0.0.0
     */
    @SerializedName("tag_name")
    private final String version;

    /** Change Log */
    @SerializedName("body")
    private final String log;

    /** Assets containing download url */
    @SerializedName("assets")
    private final List<Assets> assets;

    /**
     * Release constructor
     *
     * @param version version of latest release
     * @param log     log of latest release
     * @param assets  assets of latest release
     */
    public Release(String version, String log, List<Assets> assets) {
        this.version = version;
        this.log = log;
        this.assets = assets;
    }

    /**
     * Get latest release version
     *
     * @return latest release version
     */
    public String getVersion() {
        return version;
    }

    /**
     * Get change log of latest release
     *
     * @return change log of latest release
     */
    public String getChangeLog() {
        return log;
    }

    /**
     * Get download link of latest release
     *
     * @return download link of latest release
     */
    public String getDownloadLink() {
        return assets.get(0).getDownloadLink();
    }

    /**
     * Assets class containing download url
     */
    class Assets {
        @SerializedName("browser_download_url")
        private final String download_url;


        /**
         * Assets Constructor
         *
         * @param download_url download url
         */
        @SuppressWarnings("unused") public Assets(String download_url) {
            this.download_url = download_url;
        }

        /**
         * Get download link of latest release
         *
         * @return download link of latest release
         */
        public String getDownloadLink() {
            return download_url;
        }
    }
}

File: domain/src/main/java/tachiyomi/domain/release/model/Release.kt
package tachiyomi.domain.release.model

import android.os.Build

/**
 * Contains information about the latest release.
 */
data class Release(
    val version: String,
    val info: String,
    val releaseLink: String,
    private val assets: List<String>,
) {

    /**
     * Get download link of latest release from the assets.
     * @return download link of latest release.
     */
    fun getDownloadLink(): String {
        val apkVariant = when (Build.SUPPORTED_ABIS[0]) {
            "arm64-v8a" -> "-arm64-v8a"
            "armeabi-v7a" -> "-armeabi-v7a"
            "x86" -> "-x86"
            "x86_64" -> "-x86_64"
            else -> ""
        }

        return assets.find { it.contains("tachiyomi$apkVariant-") } ?: assets[0]
    }

    /**
     * Assets class containing download url.
     */
    data class Assets(val downloadLink: String)
}
